### DISCUSSION TITLE
spring agent - spring gateway4.x

### DISCUSSION THREAD

**Discussion by chenxu1986** [apache/skywalking#12636] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

my project use spring boot 3.2， spring-could-starter-gateway version is 4.1.5
I use skywalking agent version is  9.3.0， oapserver version is 9.3.0

I put the option-plugin  “apm-spring-cloud-gateway-4.x-plugin-9.3.0.jar”  and  “apm-spring-webflux-6.x-plugin-9.3.0.jar”   into the agent dir。

Then I conducted tests and found the following issues：

In the gateway service, a request is split into two spans, as shown in the following figure. SpringCloudGateway/RoutingFilter  and /testsk/api/inner/login/loginPwd are the same request, and you can see that their trace IDs are the same, but they are displayed as two on SK UI. Upon closer inspection, it is found that their segment IDs are different. Normally, they are all on the same gateway instance, and the segmentation ID should be consistent。

![image](https://github.com/user-attachments/assets/37945d82-50c7-4362-be28-043f3c1bea35)

![image](https://github.com/user-attachments/assets/2da93b82-bc6a-40ad-8845-c2d2aeb54c00)

The downstream service user, traceid of the gateway is completely different from that of the gateway, and has not received the traceid passed down by the gateway

![image](https://github.com/user-attachments/assets/7d1925fe-ebd4-4d23-a1e8-9f4d2b9d2de4)

So, my problem is that there is a problem with the trace passing inside the gateway. I traced the source code of the agent and found that in NettyRoutingFilterInterceptor, the afterMethod was executed before the subsequent call was initiated, causing ContextManager. stopSpan();. After exiting span, FHIR FinalizerSendInterceptor was executed, and ContextManager. isActive() became null, causing the traceid not to be passed down.
I don't know if I debugged it wrong, but I hope the trace information can be processed correctly within the gateway. I hope everyone can help me

### What you expected to happen

I hope the trace information can be processed correctly within the gateway. I hope everyone can help me

### How to reproduce

Describe the steps in the previous section

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Segment IDs are different as they are in different threads. This is expected case in the Java agent.

**Comment by wu-sheng** [comment]
I don't have any spring gateway experience, I have repeated thus many times in many issues.

It is ok you keep this public here, wait for someone could help, but I am not capable to do that.

**Comment by wu-sheng** [comment]
There are gateway tesr scenario in the code bases. You could find out how it is tested, and whether your cases and versions are covered. Someone reported weeks ago, new gateway may have different behavior, if so, we rely on new patch from someone to fix this.

**Reply by chenbihao** [reply]
> There are gateway tesr scenario in the code bases. You could find out how it is tested, and whether your cases and versions are covered. Someone reported weeks ago, new gateway may have different behavior, if so, we rely on new patch from someone to fix this.

I found a lot of information today

I'm not sure if my understanding is correct, but currently there are known issues:
-The span is fragmented and cannot be aggregated into one link
-The TID log cannot be transmitted and cannot be aggregated into one link

I have an additional question to ask. Where can I find the location of the `apm-toolkit-logback` repository? I would like to see the source code

**Reply by wu-sheng** [reply]
They are all in skywalking-java repo.

**Comment by huicunjun** [comment]
@chenxu1986 
Hello, I have encountered the same problem. Have you solved it?

**Reply by yqw570994511** [reply]
> @chenxu1986 Hello, I have encountered the same problem. Have you solved it?

Could you provide a test code for me to try and help you

**Reply by huicunjun** [reply]
> > @chenxu1986 Hello, I have encountered the same problem. Have you solved it?
> 
> Could you provide a test code for me to try and help you

gateway4.1.5，lb to SpringBoot 3.3.4 application，The link will be lost，traceid is not strung together

**Reply by huicunjun** [reply]
> > @chenxu1986 Hello, I have encountered the same problem. Have you solved it?
> 
> Could you provide a test code for me to try and help you

The local environment is easy to test, just like the main question, because the gateway link was interrupted prematurely, two traceids were generated, and the downstream service regenerated one traceid

**Reply by chenbihao** [reply]
> > @chenxu1986 Hello, I have encountered the same problem. Have you solved it?
> 
> Could you provide a test code for me to try and help you

maybe you can try this repo: [gateway demo](https://github.com/chenbihao/couponDemo/tree/main/ch6-Spring%20Cloud%20%E9%AB%98%E7%BA%A7%EF%BC%88Gateway%20%E5%BE%AE%E6%9C%8D%E5%8A%A1%E7%BD%91%E5%85%B3%EF%BC%89/middleware)

**Reply by yqw570994511** [reply]
> > > @chenxu1986 Hello, I have encountered the same problem. Have you solved it?
> > 
> > 
> > Could you provide a test code for me to try and help you
> 
> The local environment is easy to test, just like the main question, because the gateway link was interrupted prematurely, two traceids were generated, and the downstream service regenerated one traceid

This issue began to occur with gateway versions >= 4.1.2. Currently, I have not found a good way to handle this issue

**Comment by yqw570994511** [comment]
I have found a solution to address the tracking issue in versions above 4.1.2.

Use gateway4.1.5，lb to SpringBoot 3.3.4 application

The result is as follows:
<img width="416" alt="image" src="https://github.com/user-attachments/assets/abad4062-f90d-473f-97f7-932b24f7b218" />

**Comment by H2hong** [comment]
Is there no other solution?
I need to downgrade the version to gateway4.1.1+springboot3.2.12 for it to work properly

**Reply by yqw570994511** [reply, marked as answer]
To solve this problem, I plan to provide the following solutions by adding a new plugin: apm-spring-cloud-gateway-4.1.2.x-plugin ,see the link:https://github.com/apache/skywalking/discussions/12925

**Reply by chenbihao** [reply]
> To solve this problem, I plan to provide the following solutions by adding a new plugin: apm-spring-cloud-gateway-4.1.2.x-plugin ,see the link:#12925 

Does this mean that version 4.1.2 and above requires the use of this 4.1.2.x-plugin to solve this problem?

**Reply by wu-sheng** [reply]
There isn't this new thing. We just discussed about how to do that.

We need to wait for that pull request.

**Comment by zcmingg** [comment]
I also encountered this problem. I downgraded gateway4.x to 3.x to solve two span problems, but the gateway and service trace are still different (cannot be connected). The version of springboot I use is 3.3.1. Do you have any help from a boss?

**Reply by wu-sheng** [reply]
We should have a new pull request according to https://github.com/apache/skywalking/discussions/12925
